src/main/java/uk/gov/hmcts/reform/unspec/handler/callback/ConfirmServiceCallbackHandler.java

```diff
@@ -73,44 +86,64 @@ private CallbackResponse checkServedDocumentsOtherHasWhiteSpace(CallbackParams c
         }
 
         return AboutToStartOrSubmitCallbackResponse.builder()
-                   .data(data)
-                   .errors(errors)
-                   .build();
+            .data(data)
+            .errors(errors)
+            .build();
     }
 
-    private CallbackResponse addResponseDatesToCase(CallbackParams callbackParams) {
-        Map<String, Object> data = callbackParams.getRequest().getCaseDetails().getData();
-        CaseData caseData = mapper.convertValue(data, CaseData.class);
+    private CallbackResponse prepareCertificateOfService(CallbackParams callbackParams) {
+        CaseData caseData = caseDetailsConverter.toCaseData(callbackParams.getRequest().getCaseDetails());
 
         ServiceMethod serviceMethod = caseData.getServiceMethod();
-
+        Map<String, Object> data = callbackParams.getRequest().getCaseDetails().getData();
         // TODO: this field will be different (date / date time) in CCD depending on service method.
         LocalDate serviceDate = caseData.getServiceDate();
 
         LocalDate deemedDateOfService = serviceMethod.getDeemedDateOfService(serviceDate);
         LocalDateTime responseDeadline = addFourteenDays(deemedDateOfService);
-
         data.put("deemedDateOfService", deemedDateOfService);
         data.put("responseDeadline", responseDeadline);
 
+        CaseData caseDateUpdated = caseData.toBuilder()
+            .deemedDateOfService(deemedDateOfService)
+            .responseDeadline(responseDeadline)
+            .build();
```

Comment:
hmm this updating data map and updating case data seems messy. Not sure how we could improve this but it's a little confusing to need to update two objects. 

What if we passed a map to certificateOfServiceGenerator and mapped it to case data in there?

Reply:
just a thought, not saying that's what we should do 😅

Reply:
If we don't want to save the labels etc in the database, then we can remove this duplication easily by updating in CaseData object and return the same object converted to Map for the callback response. I remember there was a voice to save everything including the labels.